pennylane/core/pytree/partial.py

```diff
@@ -0,0 +1,53 @@
+import sys
+
+from functools import partial
+
+
+class HashablePartial(partial):
```

Comment:
This is taken verbatim from my package netket, where it has served us well over the last years (I wrote it myself).

It is equivalent to `functools.partial` but correctly computes its hash using the hash of its argument, contrary to `functors.partial` that would change hash every time you reconstruct it.

This is necessary to fix the bug.

@albi3ro I don't know where to insert such a utility class in the PennyLane structure, so I put it there. Is there somewhere better where it should live?

Reply:
So it seems like right now its just used by the pulse module and the jax device (which depends on the pulse module). So I would recommend keeping it close to the code that uses it for now.

This does seem like something that will be useful in the future in other locations, but I wouldn't prematurely optimize its location right now. 

So just sticking it in `pulse/parametrized_evolution.py` as an implementation detail.

Reply:
done